corehq/sql_db/shard_data_management.py

```diff
@@ -99,6 +99,54 @@ def get_count_of_unmatched_models_by_shard(database, model):
     return results
 
 
+def delete_unmatched_shard_data(database, model, query_size=5000):
+    """
+    Deletes any `model` data residing on shards that aren't expected to live in `database`.
+
+    The deletes are batched in query_size and sorted by the model's partition_attr.
+
+    Returns a generator where each value is a tuple of last value included in the query and number of records deleted
+    """
+    column_name = model.partition_attr
+    shard_id_function = _get_shard_id_function(model)
+
+    # Intentionally uses BETWEEN because it is inclusive of both values
+    query = """
+    DELETE FROM {table_name}
+    WHERE NOT ({shard_id_function}) = ANY(%s) AND {column_name} BETWEEN %s AND %s;
+    """.format(
+        shard_id_function=shard_id_function,
+        table_name=model._meta.db_table,
+        column_name=column_name
+    )
+
+    valid_shards = partition_config.get_shards_on_db(database)
+    qs = model.objects.using(database).order_by(column_name).values_list(column_name, flat=True)
+    prev_value = None
+    value = qs.first()
+    last_value = qs.order_by('-{}'.format(column_name)).first()
+    filter_expression = {}
+
+    first_run = True
+    while first_run or value < last_value:
```

Comment:
I don't fully understand the need for `last_value`. Surely we want to iterate until there is no more data?

Reply:
To be honest, I haven't htought about it a ton. This query is heavily inspired from https://github.com/dimagi/commcare-hq/blob/5a4973776118e8708ab602ef9de8a0fc7be45410/corehq/sql_db/util.py#L95-L103

Reply:
I think I have a solution to not use the extra query. will test it out and make a spearate PR for the other query

Reply:
https://github.com/dimagi/commcare-hq/pull/24763/commits/6b506cbcc5865fbb9a6944a65dcc532056a275e7 & https://github.com/dimagi/commcare-hq/pull/24819